crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java

```diff
@@ -0,0 +1,143 @@
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
+
+/**
+ * Serializer for wrapping and unwrapping {@link KeyMaterial}. The {@link #wrap} method returns the KeyMaterial
+ * serialized as follows, which is the same format the {@link #unwrap} method expects:
+ *
+ * <pre>
+ *  +--------------------------------------------------------------+
+ *  | version | wrapping iv | key algorithm length | key algorithm |
+ *  |   byte  |    byte[]   |          int         |     byte[]    |
+ *  +--------------------------------------------------------------+
+ *  +-------------------------------------------------------+
+ *  | wrapped key length | wrapped key | iv length |   iv   |
+ *  |        int         |    byte[]   |    int    | byte[] |
+ *  +-------------------------------------------------------+
+ * </pre>
+ */
+enum SymmetricKeySerializerV3 implements SymmetricKeySerializer {
+    INSTANCE;
+
+    private static final String AES_CTR_NO_PADDING = "AES/CTR/NoPadding";
+    private static final int IV_SIZE = 16;
+
+    static final int VERSION = 3;
+
+    @Override
+    public byte[] wrap(KeyMaterial keyMaterial, SecretKey key) {
+        ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
+        DataOutputStream stream = new DataOutputStream(byteStream);
+
+        byte[] wrappingIv = KeyMaterials.generateIv(IV_SIZE);
+        Cipher keyWrappingCipher = getCipher(Cipher.WRAP_MODE, key, wrappingIv);
+        SecretKey secretKey = keyMaterial.getSecretKey();
+
+        try {
+            stream.write(VERSION);
+
+            stream.write(wrappingIv);
+
+            String keyAlgorithm = secretKey.getAlgorithm();
+            stream.writeInt(keyAlgorithm.length());
+            stream.write(keyAlgorithm.getBytes(StandardCharsets.UTF_8));
+
+            byte[] encryptedKey = keyWrappingCipher.wrap(secretKey);
+            stream.writeInt(encryptedKey.length);
+            stream.write(encryptedKey);
+
+            byte[] iv = keyMaterial.getIv();
+            stream.writeInt(iv.length);
+            stream.write(iv);
+
+            stream.close();
+            return byteStream.toByteArray();
+        } catch (IOException | InvalidKeyException | IllegalBlockSizeException e) {
+            throw new RuntimeException("Unable to wrap key", e);
+        }
+    }
+
+    @Override
+    public KeyMaterial unwrap(byte[] wrappedKeyMaterial, SecretKey key) {
+        DataInputStream stream = new DataInputStream(new ByteArrayInputStream(wrappedKeyMaterial));
+
+        try {
+            int version = stream.read();
+            Preconditions.checkArgument(VERSION == version,
+                    "Invalid serialization format version. Expected %s but found %s", VERSION, version);
+
+            byte[] wrappingIv = new byte[IV_SIZE];
+            stream.read(wrappingIv);
```

Comment:
This needs to be `readFully` as `read` may not read all bytes

Reply:
thanks. @uschi2000 should probably have an error-prone check for this.

Reply:
What would you want to check? That you have to use the return value of read?

Reply:
Yep, IntelliJ's static analysis flags this, see also https://wiki.sei.cmu.edu/confluence/display/java/FIO10-J.+Ensure+the+array+is+filled+when+using+read%28%29+to+fill+an+array

This specific case would likely never hit an issue since we're reading from ByteArrayInputStream, but if someone copy/pasted to read from network or disk, sad times may ensue.